### PR TITLE
LPD-26381 Fix behavior of ERC in service-builder for versioned entities

### PR DESCRIPTION
**Pull request by albertojml**
**What's changed?:**
- `persistence_impl.ftl` has been adapted to the behavior of versioned entities so that the external reference code type fields are compatible.
- `persistence_test.ftl` has new tests to check some edge-cases for the ERC + Versioned entities behavior.
- Added new test entity in the `service-builder-test` service for checking the new ERC + Versioned Entity behavior.
- Executed the Build Service task, modifying the `fragment-service` module in the process.

**_IMPORTANT NOTE:_** This PR needs to be reviewed also by @liferay-echo

See the following **Jira Ticket**: [LPD-26381](https://liferay.atlassian.net/browse/LPD-26381)

### CODE REVIEW

**Comment by albertojml** [comment]
ci:test:relevant

**Review by 4lejandrito — CHANGES_REQUESTED** [review]
@albertojml I think we need tests for this. Service builder [autogenerates tests](https://github.com/liferay/liferay-portal/blob/c41994b5fb6f84da38c09e4f74055d97a0a39e96/modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl) like rest builder. It would be great if we could cover the semantics of ERC + Versioning with some automated tests.

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

**Comment by albertojml** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
I think in the description the name of the file is not correct: `persistence_test.ftl` vs `persistance-test.ftl`

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

File: modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl [hunk 1]
```diff
@@ -330,6 +330,222 @@ public class ${entity.name}PersistenceTest {
...
+						);
+					</#if>
+				</#list>
+
+				_${entity.pluralVariableName}.add(_persistence.update(${entity.variableName}2));
+			}
+
+			protected ${entity.name} addDraftForExisting${entity.name}(${entity.name} existing${entity.name}) throws Exception {
+				<#if entity.hasCompoundPK()>
+					${entity.PKClassName} pk = new ${entity.PKClassName}(
```
Comment (ccorreagg) [inline modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:484]: Are we testing this code? I think with the current service builder modules in Liferay, it is not possible to have this use case

**Review by ccorreagg — CHANGES_REQUESTED** [review]
Please review my comments.

BTW, I have seen a couple of recommendations related the commits:

1) You use sometimes different tenses when creating the commits:

`LPD-26381 Add external reference code to FragmentEntry`
`LPD-26381 Added new service-builder test entity for checking the new ERC for versioned entities behavior`

Maybe better to unify, for example, in present tense imperative (`Added new service-builder test` --> `Add new service-builder test`)

2) The author seems to be your git username, not your name (`Author: albertojml <alberto.moreno@liferay.com>`) instead of `Author: Alberto Moreno <alberto.moreno@liferay.com>`, not sure if this is mandatory, but easier to identify the author

File: modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/service/persistence/test/FragmentEntryPersistenceTest.java [hunk 2]
```diff
@@ -831,6 +1119,17 @@ private void _assertOriginalValues(FragmentEntry fragmentEntry) {
 				fragmentEntry, "getColumnOriginalValue",
 				new Class<?>[] {String.class}, "fragmentEntryKey"));
 
+		Assert.assertEquals(
+			fragmentEntry.getExternalReferenceCode(),
+			ReflectionTestUtil.invoke(
+				fragmentEntry, "getColumnOriginalValue",
+				new Class<?>[] {String.class}, "externalReferenceCode"));
+		Assert.assertEquals(
```
[the remark this answers is not in the document]
Comment (albertojml) [reply modules/apps/fragment/fragment-test/src/testIntegration/java/com/liferay/fragment/service/persistence/test/FragmentEntryPersistenceTest.java:1127]: I'm seing that this is something that all service builder entities do. See for example the [Blogs](https://github.com/liferay/liferay-portal/blob/6ce0b9eb049e99a50285a324a3bd4afd76d340ab/modules/apps/blogs/blogs-test/src/testIntegration/java/com/liferay/blogs/service/persistence/test/BlogsEntryPersistenceTest.java#L851) and persistence tests.

File: modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl [hunk 3]
```diff
@@ -1399,7 +1399,9 @@ public class ${entity.name}PersistenceTest {
 
 				${entity.variableName}.set${entityColumn.methodName}(
 
-				<#if stringUtil.equals(entityColumn.type, "boolean")>
+				<#if stringUtil.equals(entityColumn.name, "headId") && entity.versionEntity??>
+					- RandomTestUtil.nextLong()
```
[the remark this answers is not in the document]
Comment (albertojml) [reply modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:1403]: The thing is that since we have started testing something more specific to the versioned entities context, the negative headId value indicates that is the head of an specific entry.

If we don't change that value, we would be treating as "Head" a draft entity and will be confusing during testing.

[the remark this answers is not in the document]
Comment (ccorreagg) [reply modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:1403]: Yes, what I mean is, should the value be the negative id of the head?

File: modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl [hunk 4]
```diff
@@ -330,6 +330,222 @@ public class ${entity.name}PersistenceTest {
...
+						<#else>
+							Assert.assertEquals(${entity.variableName}.get${entityColumn.methodName}(), draft${entity.name}.get${entityColumn.methodName}());
+						</#if>
+					</#if>
+				</#list>
+
+			}
+
+			@Test(expected = ${duplicateEntityExternalReferenceCode}Exception.class)
+			public void testUpdateAddWithExistingExternalReferenceCodeHead()
```
[the remark this answers is not in the document]
Comment (albertojml) [reply modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:380]: What do you mean exactly with this? 

The tests that have been added cover the following:
- You can't create a new FragmentEntry with an already existing ERC.
- You can create a new DraftFragmentEntry with an existing ERC for an already existing entry.
- You can't create a new DraftFragmentEntry if there's already another DraftFragmentEntry with the same ERC.

I know that because of the way it's implemented, the last case always works if the first one also works but I've added it just in case this implementation changes in the future but I don't know if this is what you are trying to say...

File: modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl (hunk 1 above)
Comment (albertojml) [reply modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:484]: Yes, you're right. In fact Service Builder checks specifically that versioned entities don't have compound PKs so I will delete this part of the code.

File: modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ccorreagg) [reply modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:380]: As we talked, we are already testing with the same ERC and head value

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

File: modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl [hunk 5]
```diff
@@ -330,6 +330,154 @@ public class ${entity.name}PersistenceTest {
...
+						${entity.variableName}.get${entityColumn.methodName}());
+
+					</#if>
+				</#list>
+
+				_${entity.pluralVariableName}.add(_persistence.update(draft${entity.name}));
+
+				<#list entity.regularEntityColumns as entityColumn>
+					<#if !entityColumn.primary && (validator.isNull(parentPKColumn) || (parentPKColumn.name != entityColumn.name))>
+						<#if stringUtil.equals(entityColumn.name, "mvccVersion")>
```
[the remark this answers is not in the document]
Comment (ccorreagg) [reply modules/util/portal-tools-service-builder/src/main/resources/com/liferay/portal/tools/service/builder/dependencies/persistence_test.ftl:376]: The test passes without the if clause. Is ignoring the assertion needed? :thinking: 

```
com.liferay.fragment.service.persistence.test.FragmentEntryPersistenceTest > testCreateDraft STARTED

com.liferay.fragment.service.persistence.test.FragmentEntryPersistenceTest > testCreateDraft PASSED
```

**Comment by ealonso** [comment]
Hey @albertojml,

I've updated our pr [here](https://github.com/liferay-echo/liferay-portal/pull/15752) and I've added some tests to check the new functionality.

One of those tests is related with adding a duplicate ERC, in other cases without versioning when we are adding a duplicate ERC we are getting a Duplicate...ExternalReferenceCodeException but in this case we are getting javax.persistence.PersistenceException. See the stacktrace:

```
com.liferay.style.book.service.test.StyleBookEntryLocalServiceTest > testAddStyleBookEntryWithExistingExternalReferenceCode FAILED
    java.lang.Exception: Unexpected exception, expected<com.liferay.style.book.exception.DuplicateStyleBookEntryExternalReferenceCodeException> but was<javax.persistence.PersistenceException>
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable._processExpectedException(TestExecutorRunnable.java:407)
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable.access$400(TestExecutorRunnable.java:57)
    	at com.liferay.arquillian.extension.junit.bridge.server.TestExecutorRunnable$3.evaluate(TestExecutorRunnable.java:371)
    	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
```

Is it the expected behavior?

**Comment by albertojml** [comment]
ci:test:sf

**Comment by albertojml** [comment]
ci:test:relevant

**Comment by albertojml** [comment]
ci:test:relevant

**Review by 4lejandrito — APPROVED** [review]

**Comment by 4lejandrito** [comment]
ci:forward:force
